### PR TITLE
cmake: Rework `WITH_CCACHE`

### PR DESCRIPTION
**Pull request by hebasto**
This PR makes `WITH_CCACHE` an opportunistic boolean option with the default value `ON`.

Also the build system learned to tell ccache in the [masquerade](https://ccache.dev/manual/4.9.1.html#_run_modes) mode, which addresses this https://github.com/hebasto/bitcoin/pull/93#issuecomment-1929663187.

The `TristateOption` module has been deleted as planned.

### CODE REVIEW

File: cmake/ccache.cmake [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+if(MSVC AND WITH_CCACHE)
+  if(CMAKE_VERSION VERSION_LESS 3.24)
+    # ccache >= 4.8 requires compile batching turned off that is available since CMake 3.24.
+    # See https://github.com/ccache/ccache/wiki/MS-Visual-Studio
+    set(WITH_CCACHE "OFF, CMake >= 3.24 is required")
+  else()
+    set(ccache_hints)
+    if(EXISTS "$ENV{ChocolateyInstall}")
+      # Bypass a shim executable provided by Chocolatey.
```
Comment (fanquake) [inline cmake/ccache.cmake:13]: Is this not something that users of this package manger should be fixing on their own machine? In any case, the comment doesn't explain why it's needed (same goes for all other comments here).

**Comment by hebasto** [comment]
Rebased.

File: cmake/ccache.cmake (hunk 1 above)
Comment (hebasto) [reply cmake/ccache.cmake]: > Is this not something that users of this package manger should be fixing on their own machine?

That's indeed unfortunate that Chocalately's `ccache.exe` shim executable can't be moved or copied without being broken. But it is our intention to copy it in order to integrate with MSVC, therefore, I believe that this is our responsibility to provide a workaround for such a case. Besides, it is only a single path hint.

Of course, this workaround can be removed in the future if it leads to extra maintenance burden.

> In any case, the comment doesn't explain why it's needed (same goes for all other comments here).

I’ve reworked the comments in the hope of addressing your comment.

Comment (fanquake) [reply cmake/ccache.cmake]: > But it is our intention to copy it in order to integrate with MSVC.

I still don't understand why we even need to do this though, to make ccache "work".

Comment (hebasto) [reply cmake/ccache.cmake]: [Reworked](https://github.com/hebasto/bitcoin/pull/215#issuecomment-2173378806).

**Comment by hebasto** [comment]
https://github.com/hebasto/bitcoin/pull/215#discussion_r1642514705:
> I thought we were going to remove all of this for now?

Sure. MSVC support has been removed.

File: cmake/ccache.cmake [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  # https://github.com/ccache/ccache/wiki/MS-Visual-Studio
+  find_program(CCACHE_EXECUTABLE ccache)
+  if(CCACHE_EXECUTABLE)
+    file(COPY_FILE ${CCACHE_EXECUTABLE} ${CMAKE_BINARY_DIR}/cl.exe ONLY_IF_DIFFERENT)
+    list(APPEND CMAKE_VS_GLOBALS
+      "CLToolExe=cl.exe"
+      "CLToolPath=${CMAKE_BINARY_DIR}"
+      "DebugInformationFormat=OldStyle"
+    )
+    set(CMAKE_VS_NO_COMPILE_BATCHING ON)
```
[the remark this answers is not in the document]
Comment (hebasto) [reply cmake/ccache.cmake]: Parallelism is provided by means of `"UseMultiToolTask=true"`.

File: cmake/ccache.cmake [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2023-present The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or https://opensource.org/license/mit/.
+
+if(MSVC AND WITH_CCACHE)
+  # For integration of ccache and MSVC, please refer to
```
[the remark this answers is not in the document]
Comment (hebasto) [reply cmake/ccache.cmake]: [Reworked](https://github.com/hebasto/bitcoin/pull/215#issuecomment-2173439295).

**Comment by m3dwards** [comment]
ACK https://github.com/hebasto/bitcoin/pull/215/commits/4cd15f1c64537fc9964dcdb7ff763884ff4e9278

Tested on Debian and saw ccache running and second builds were much faster. I also uninstalled ccache and configure correctly configured it to be off. I then tried to force ccache on and configure would not allow it to be on without ccache installed.

**Pull request merged**